substrate/dex/pallet/src/lib.rs

```diff
@@ -161,29 +161,75 @@ pub mod pallet {
-      let mut iter = OraclePrices::<T>::iter_key_prefix(coin);
-      // the first key will be the lowest price due to the keys being lexicographically ordered.
-      iter.next().map(|amount| Amount(u64::from_be_bytes(amount)))
-    }
-  }
+
+  /// Same as `OraclePrices` but in reverse order, since the keys we save to this map are
+  /// lexiogracpilly reversed. Hence this map stores the amounts from hight to low.
+  #[pallet::storage]
+  pub type OraclePricesReverse<T: Config> =
+    StorageDoubleMap<_, Identity, Coin, Identity, [u8; 8], u16, OptionQuery>;
 
   #[pallet::storage]
   #[pallet::getter(fn oracle_value)]
   pub type OracleValue<T: Config> = StorageMap<_, Identity, Coin, Amount, OptionQuery>;
 
+  /// Current median price of the prices in the oracle prices maps at any given time.
+  #[pallet::storage]
+  #[pallet::getter(fn median_price)]
+  pub type MedianPrice<T: Config> = StorageMap<_, Identity, Coin, Amount, OptionQuery>;
+  impl<T: Config> Pallet<T> {
+    /// Update the current median according to new state of oracle prices.
+    pub fn update_median_price(
+      coin: &Coin,
+      inserted: u64,
+      removed: Option<u64>,
+      n: BlockNumberFor<T>,
+    ) {
+      let current = Self::median_price(coin).unwrap_or(Amount(0)).0;
+
+      // if we didn't remove  an amount from the window, just means we still didn't hit
+      // the full size of the window.
+      let next = if removed.is_none() {
+        if n % 2u32.into() == 1u32.into() {
+          if inserted >= current {
+            OraclePrices::<T>::iter_key_prefix_from(coin, current.to_be_bytes().to_vec()).next()
+          } else {
+            let reversed = reverse_lexicographic_order(current.to_be_bytes());
+            OraclePricesReverse::<T>::iter_key_prefix_from(coin, reversed.to_vec())
+              .next()
+              .map(reverse_lexicographic_order)
+          }
+        } else {
+          Some(current.to_be_bytes())
+        }
+      } else {
+        let removed = removed.unwrap();
+        // if inserted is higher and removed is lower, we just need to go to the next key on the
+        // low to high map to find the current median.
+        if inserted > current && removed <= current {
+          OraclePrices::<T>::iter_key_prefix_from(coin, current.to_be_bytes().to_vec()).next()
+        } else if inserted < current && removed >= current {
+          // if inserted is lower and removed is higher, we go to next key on the reverse map.
+          let reversed = reverse_lexicographic_order(current.to_be_bytes());
+          OraclePricesReverse::<T>::iter_key_prefix_from(coin, reversed.to_vec())
+            .next()
+            .map(reverse_lexicographic_order)
+        } else {
+          // no change if inserted and removed both in same side(both smaller or bigger).
+          Some(current.to_be_bytes())
+        }
+      };
+
+      // TODO: use current or 0 for unwrap?
+      let next = next.unwrap_or(current.to_be_bytes());
```

Comment:
Is there a way this can fail? AFAICT, it'd require current be the edge and we try to further move. That'd only fail if there's a single item? We can short-circuit this algorithm while length < 3 for such cases.

The failure point of this median algorithm is how any drift is never corrected and always accumulates. We definitely shouldn't flatline to 0 accordingly, yet remaining on current sounds like it'd introduce a drift.